### PR TITLE
ffi: Add support for parsing wildcard queries using CLP's default variable patterns.

### PR DESCRIPTION
**Pull request by kirkrodrigues**
# Description

This adds support for turning a wildcard query into a series of subqueries that can be run against CLP-encoded messages. For more details about the feature, see components/core/src/ffi/search/README.md

# Validation performed

Tested with new unit tests.

### CODE REVIEW

**Comment by kirkrodrigues** [comment]
@haiqi96, would you be able to review this?

**Review by haiqi96 — COMMENTED** [review]
First bunch of comments

File: components/core/src/ffi/search/WildcardToken.cpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        // dictionary variable
+        m_possible_variable_types.push_back(TokenType::DictionaryVariable);
+
+        m_current_interpretation_idx = 0;
+    }
+
+    template <typename encoded_variable_t>
+    bool WildcardToken<encoded_variable_t>::add_to_logtype_query (string& logtype_query) const {
+        auto current_interpretation = m_possible_variable_types[m_current_interpretation_idx];
+        if (TokenType::StaticText == current_interpretation) {
```
Comment (haiqi96) [inline components/core/src/ffi/search/WildcardToken.cpp:166]: Quick question: why we don't need to consider the case for begin = wildcard?

**Review by haiqi96 — COMMENTED** [review]
another comment

File: components/core/src/ffi/search/CompositeWildcardToken.cpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+        const QueryWildcard* last_wildcard = nullptr;
+        bool wildcardInVariable = false;
+        size_t var_begin_pos, var_end_pos;
+        for (const auto& w : m_wildcards) {
+            switch (w.get_current_interpretation()) {
+                case WildcardInterpretation::NoDelimiters:
+                    wildcardInVariable = true;
+                    break;
+                case WildcardInterpretation::ContainsDelimiters: {
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/CompositeWildcardToken.cpp:173]: Modified the function header comment to try and make it more clear what the steps are here.

File: components/core/src/ffi/search/CompositeWildcardToken.cpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                    // Determine var_end_pos
+                    if (WildcardType::ZeroOrMoreChars == w.get_type()) {
+                        // Include the wildcard in the token
+                        var_end_pos = wildcard_pos + 1;
+                        wildcardInVariable = true;
+                    } else {
+                        var_end_pos = wildcard_pos;
+                    }
+
+                    if (wildcardInVariable) {
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/CompositeWildcardToken.cpp]: Good catch.

File: components/core/src/ffi/search/README.md [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+There are four query interpretations for the example:
+
+1. Interpretation 1:
+   * Dictionary variable queries: `["*task*"]`
+   * Encoded variable queries: `["0.3*"]`
+   * Logtype query: `*<dict-var> took <float>*`
+2. Interpretation 2:
+   * Dictionary variables queries: `["*task*", "0.3*"]`
+   * Encoded variable queries: `[]`
+   * Logtype query: `*<dict-var> took <dict-var>*`
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/README.md]: Good catch.

File: components/core/src/ffi/search/README.md [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+`container`.
+
+Comparing the first and third interpretation in Table 1, we can see that the
+third is a more general version of the first. As a result, we don't need to
+consider the first interpretation. We can generalize this as follows: *If a `*`
+is interpreted to have a different type than either of the characters
+surrounding it, the tokenization should split the string at the `*` while
+leaving a `*` attached to the surrounding characters.*
+
+So the wildcard-containing token, `*container*to`, can be tokenized either as:
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/README.md]: Good catch.

File: components/core/src/ffi/search/WildcardToken.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+                ++num_negative_signs;
+                if (num_negative_signs > 1) {
+                    // Contains multiple negative signs
+                    return false;
+                }
+            } else if ('0' <= c && c <= '9') {
+                ++num_digits;
+                constexpr size_t cMaxDigitsInRepresentableIntVar =
+                        std::is_same_v<encoded_variable_t, four_byte_encoded_variable_t>
+                        ? 10 : 19;
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/WildcardToken.cpp]: Good idea.

File: components/core/src/ffi/search/WildcardToken.cpp (hunk 1 above)
Comment (kirkrodrigues) [reply components/core/src/ffi/search/WildcardToken.cpp:177]: Added more comments to clarify.

File: components/core/src/ffi/search/query_methods.cpp [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+#include "CompositeWildcardToken.hpp"
+#include "QueryMethodFailed.hpp"
+
+using std::pair;
+using std::string;
+using std::string_view;
+using std::variant;
+using std::vector;
+
+auto TokenGetBeginPos = [] (const auto& token) { return token.get_begin_pos(); };
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/query_methods.cpp]: * They didn't need to be inside the namespace, but following the philosophy of keeping things closer to their use, I moved them inside the namespace.
* We're using the methods with `std::variant` which requires a visitor lambda/operators to be able to call a method on one of the types inside the variant. Unless there's some syntax I'm missing where I could just call `token.get_begin_pos()`?

File: components/core/src/ffi/search/query_methods.cpp [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+        vector<variant<ExactVariableToken<encoded_variable_t>,
+                CompositeWildcardToken<encoded_variable_t>>> tokens;
+        vector<CompositeWildcardToken<encoded_variable_t>*> composite_wildcard_tokens;
+        tokenize_query(wildcard_query, tokens, composite_wildcard_tokens);
+
+        string logtype_query;
+        vector<variant<ExactVariableToken<encoded_variable_t>,
+                WildcardToken<encoded_variable_t>>> query_vars;
+        while (true) {
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/query_methods.cpp]: Good idea.

File: components/core/src/ffi/search/query_methods.cpp [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+            vector<CompositeWildcardToken<encoded_variable_t>*>& composite_wildcard_tokens
+    ) {
+        // Tokenize query using delimiters to get definite variables and tokens
+        // containing wildcards (potential variables)
+        size_t end_pos = 0;
+        auto wildcard_query_length = wildcard_query.length();
+        while (true) {
+            auto begin_pos = end_pos;
+
+            // Find next wildcard or delimiter
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/query_methods.cpp]: Good catch.

File: components/core/src/ffi/search/query_methods.cpp [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+            bool is_escaped = false;
+            bool contains_wildcard = false;
+            for (; begin_pos < wildcard_query_length; ++begin_pos) {
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/query_methods.cpp]: Good idea.

File: components/core/src/ffi/search/query_methods.cpp [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+                break;
+            }
+
+            bool contains_decimal_digit = false;
+            bool contains_alphabet = false;
+
+            // Find next delimiter
+            is_escaped = false;
+            end_pos = begin_pos;
+            while (end_pos < wildcard_query_length) {
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/query_methods.cpp]: Good idea.

File: components/core/src/ffi/search/query_methods.cpp [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+                }
+            } else {
+                string_view variable(wildcard_query.cbegin() + begin_pos, end_pos - begin_pos);
+                // Treat token as variable if:
+                // - it contains a decimal digit, or
+                // - it's directly preceded by an equals sign and contains an
+                //   alphabet, or
+                // - it could be a multi-digit hex value
+                if (contains_decimal_digit ||
+                    (begin_pos > 0 && '=' == wildcard_query[begin_pos - 1] && contains_alphabet) ||
```
[the remark this answers is not in the document]
Comment (kirkrodrigues) [reply components/core/src/ffi/search/query_methods.cpp:183]: Something like `=-;`. `-` is not a delimiter, but it appears between two delimiters `=` and `;`, and is preceded by an `=`.

**Review by haiqi96 — APPROVED** [review]
Approving the PR since all comments are addressed.

**Review by vishalpalaniappan — APPROVED** [review]
Rubber stamp approval.

**Pull request merged**